Document `autoLockOnActor`

fig02 opened the pull request:
"Auto lock-on actor" sets lock-on without players input. Its used twice in the whole game.

Here is a video demonstration of the quirk mentioned in the redead actor. I think its worthy of a bug tag, but opted not to put it for now. Curious on opinions.
https://github.com/user-attachments/assets/7e2d04be-70e8-4293-9cfb-f221ccc88f50

cadmic commented on src/overlays/actors/ovl_En_Rd/z_en_rd.c:
```diff
@@ -606,7 +615,9 @@ void EnRd_AttemptPlayerFreeze(EnRd* this, PlayState* play) {
             Rumble_Request(this->actor.xzDistToPlayer, 255, 20, 150);
-            func_8008EEAC(play, &this->actor);
+
+            // The same note mentioned with this function call in `EnRd_WalkToPlayer applies here too
```
only one ` here

cadmic commented on src/code/z_player_lib.c:
```diff
@@ -742,12 +742,17 @@ void func_8008EE08(Player* this) {
+/**
+ * Sets the "auto lock-on actor" to lock onto an actor without Player's input.
+ * This function will first release any existing lock-on or (try to) release parallel.
+ * The new lock-on is considered "friendly" even if the actor is actually hostile.
```
is it intended that it uses "friendly" lock-on? what effects does this have?

fig02 replied:
Well for one player wont have his "hostile" battle response. but this doesnt matter in both contexts this is used anyway cause he is either frozen or in a chokehold.

I think more importantly it will use the "focus" or "keepon" camera mode instead of "battle". Which just leads to a slightly different looking camera.

cadmic replied:
will the camera switch back to being "hostile" after the chokehold is released? or is it permanently friendly?

fig02 replied:
So it seems like with switch targeting you cannot keep the lockon held after an auto lockon ends. It will release automatically. This is likely because `PLAYER_STATE2_LOCK_ON_WITH_SWITCH` never gets set.

You can keep the lock-on held with hold targeting however. And it indeed switches from "friendly" to "hostile" when the auto lockon is over.

See clip below with the camera debugger enabled. "KEEPON" is friendly actor focus and "BATTLE" is hostile.

https://github.com/user-attachments/assets/fd8100ab-29f1-4efd-88db-3120a57b9e73

cadmic replied:
oh interesting, thanks for looking. Maybe copy-paste that into the comment?

fig02 replied:
Sure.

I think general speaking what is currently called the `CAM_MODE_Z_TARGET_FRIENDLY` mode is really just a general actor focus mode (`CAM_MODE_ACTOR_FOCUS`). Nothing it particularly "friendly" about it.

However `PLAYER_STATE1_FRIENDLY_ACTOR_FOCUS` being set would make Link unable to have his battle response, if he were actionable during that time.

Not sure how in depth to go here with the comment but I will try to make it reasonable

cadmic commented on src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -3698,15 +3699,18 @@ void func_80836BEC(Player* this, PlayState* play) {
-            } else if (this->unk_684 != NULL) {
-                this->focusActor = this->unk_684;
+            } else if (this->autoLockOnActor != NULL) {
+                // `autoLockOnActor` does not take precedence over `focusActor` if it already exists.
+                // However, `autoLockOnActor` is expected to be set with `Player_SetAutoLockOnActor`
+                //  which will release any existing lock-on. So it effectively does take precedence.
```
I'm not sure how to read this, could you reword? What is `it` in the first sentence? And since this is overriding focusActor with autoLockOnActor if autoLockOnActor exists, doesn't that mean that autoLockOnActor takes precedence here too?

fig02 replied:
"it" is `focusActor` 
if you look at the surrounding if/if-else youll see that player first prioritizes `focusActor` if it already exists. Only if it doesnt will `focusActor` get set to `autoLockOnActor`.

In other words, `autoLockOnActor` itself does not force itself to be the main focus if it is set. You need to use the function to release a lock on first in order to "force" this one to be used

cadmic replied:
> if you look at the surrounding if/elseif youll see that player first prioritizes focusActor if it already exists
I see thanks, maybe move this comment closer to here, or mention the previous if in the comment?

cadmic reviewed (commented):
very interesting

krm01 reviewed (approved):
LGTM. unsure if the redead thing deserves "bug", personally I'd say no, it's just an interesting interaction with freezeTimer.

fig02 commented:
Imo its a bit more than an "interesting interaction", a whole piece of the auto lock on is missing. And the strange 1 frame lock on that occurs when the freeze ends is weird and feels buggy

krm01 commented:
the main reason i say that is because i think it's better to not have the lock-on reticle appear, and it's such a fundamental basic part of redead interaction, it's hard to believe the current behavior wasn't desired / intended. imo it's definitely a case of "that's not a bug it's a feature" :)

fig02 commented:
Right you only hold that view because its how youre used to it working. Which I agree, it would feel weird if it was there cause of what we expect it to look like. But I still think its a bug lol

cadmic commented:
FWIW a while back I tried to add the `@bug` tag to some "not a bug it's a feature"s in camera a while ago but it wasn't popular: #1563
So it seems like the criteria for the `@bug` tag is pretty strong, it has to be an obvious bug, and not merely a coding mistake or an "oddity"

fig02 commented:
If it wasnt for the buggy behavior at the end of the auto lock on I would probably hold the same opinion as that case there. I still think that this is a true bug, but since its not a popular opinion I wont add it.

Anyway, I tried addressing the comment changes. Its hard to word this stuff while trying to be concise and also complete

cadmic reviewed (approved):
Thanks, that seems pretty good. Didn't mean to hold this up so much, sorry

fig02 commented:
No its fine, the time to write this stuff down is now, while its in my brain :D just needed a bit to figure out how I wanted to say it
